### ISSUE TITLE
'Something' used before it is defined

### ISSUE THREAD

**Issue by cwoolum** [johnpapa/angular-styleguide#345]
I have a question about Y033. I like the idea of putting the bindable members up top but this interferes with the [latedef](http://jshint.com/docs/options/#latedef) rule in jsHint. Is there a good middle road to appease the jsHint gods? We might end up placing only function assignments at the bottom. If the precedent is that functions are assigned at the end of the function, it is still readable and easy to work with.

I can create a PR with the modifications if you agree.

**Comment by wardbell** [comment]
Are you assigning a function to a variable below or defining the function  below? We recommend the latter which should not trigger that rule.

**Comment by cwoolum** [comment]
My specific example is this.... This causes the error:

```javascript
var vm = this;
vm.employee = {};
        
vm.createEmployee = createEmployee;
        
function createEmployee(employee) {
  return employeesService.$save(employee).then(function (response) {
    console.log(response);
    }).catch(function (error) {
      var errorMessage = error.data.errorDescription;
      console.log(errorMessage);
    });
}

```

I propose that instead, we do this

```javascript
// Normal assignments here
var vm = this;
vm.pageTitle = 'Add Employee';
vm.employee = {};
      
function createEmployee(employee) {
  return employeesService.$save(employee).then(function (response) {
    console.log(response);
      }).catch(function (error) {
        var errorMessage = error.data.errorDescription;
        console.log(errorMessage);
      });
}

// Function assignments down here
vm.createEmployee = createEmployee;
```

**Comment by dietergeerts** [comment]
Don't do this! The first example is much better as you can see what's on your controller/viewmodel in one glance. If you know Javascript, then you'll also know that it will 'bubble up' variable declarations by itself, so if you use the second example, it will become the first eventually.

**Comment by cwoolum** [comment]
So do you think JSHint is wrong about this? I don't want to sacrifice code quality for readability.

**Comment by gkalpak** [comment]
I don't get a warning from jshint the former.
I guess others don't get one either...

**Comment by ArtemFitiskin** [comment]
Try JSHint **latedef** option.

From [JSHint Options](http://jshint.com/docs/options/#latedef)

> This option prohibits the use of a variable before it was defined. JavaScript has function scope only and, in addition to that, all variables are always moved—or hoisted— to the top of the function. This behavior can lead to some very nasty bugs and that's why it is safer to always use variable only after they have been explicitly defined.
> Setting this option to "nofunc" will allow function declarations to be ignored.

Probably helps you.

p.s. Is it duplicate #165 issue?

**Comment by johnpapa** [comment]
JSHint isnt right or wrong, its a set of rules or hints that you can apply. There is a jshtinrc file where you can adjust the rules. You can see an example of one in the style guide.

The latedef option is one I use so functions can be defined late, but not variables. It's all in the jshintrc file

**Comment by johnpapa** [comment]
closing as dupe of https://github.com/johnpapa/angular-styleguide/issues/165

**Comment by cwoolum** [comment]
Aah awesome. Thanks!
